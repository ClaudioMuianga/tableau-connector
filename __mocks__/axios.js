import Promise from 'promise'

let mockResponse = {}

const __setMockResponse = (responseData) => {
  mockResponse.data = responseData
}

const get = () => {
  return new Promise(function (resolve) {
    resolve(mockResponse)
  })
}

export default {
  __setMockResponse,
  get
}