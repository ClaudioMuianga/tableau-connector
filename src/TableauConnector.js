import axios from 'axios'

const tableau = window.tableau

const schemaMap = {
  'http://www.w3.org/2001/XMLSchema#boolean': tableau.dataTypeEnum.bool,
  'http://www.w3.org/2001/XMLSchema#integer': tableau.dataTypeEnum.int,
  'http://www.w3.org/2001/XMLSchema#nonPositiveInteger': tableau.dataTypeEnum.int,
  'http://www.w3.org/2001/XMLSchema#nonNegativeInteger': tableau.dataTypeEnum.int,
  'http://www.w3.org/2001/XMLSchema#negativeInteger': tableau.dataTypeEnum.int,
  'http://www.w3.org/2001/XMLSchema#long': tableau.dataTypeEnum.int,
  'http://www.w3.org/2001/XMLSchema#int': tableau.dataTypeEnum.int,
  'http://www.w3.org/2001/XMLSchema#short': tableau.dataTypeEnum.int,
  'http://www.w3.org/2001/XMLSchema#byte': tableau.dataTypeEnum.int,
  'http://www.w3.org/2001/XMLSchema#unsignedLong': tableau.dataTypeEnum.int,
  'http://www.w3.org/2001/XMLSchema#unsignedInt': tableau.dataTypeEnum.int,
  'http://www.w3.org/2001/XMLSchema#unsignedShort': tableau.dataTypeEnum.int,
  'http://www.w3.org/2001/XMLSchema#unsignedByte': tableau.dataTypeEnum.int,
  'http://www.w3.org/2001/XMLSchema#positiveInteger': tableau.dataTypeEnum.int,
  'http://www.w3.org/2001/XMLSchema#decimal': tableau.dataTypeEnum.float,
  'http://www.w3.org/2001/XMLSchema#float': tableau.dataTypeEnum.float,
  'http://www.w3.org/2001/XMLSchema#double': tableau.dataTypeEnum.float,
  'http://www.w3.org/2001/XMLSchema#date': tableau.dataTypeEnum.date,
  'http://www.w3.org/2001/XMLSchema#dateTime': tableau.dataTypeEnum.datetime,
  'http://www.w3.org/2001/XMLSchema#dateTimeStamp': tableau.dataTypeEnum.datetime
}

const queryTable = 'TableColumns'

export default class TableauConnector {

  constructor () {
    this.connector = tableau.makeConnector()
    this.connector.getSchema = this.getSchema
    this.connector.getData = this.getData
    this.submit = tableau.submit
    tableau.registerConnector(this.connector)
  }

  getApiEndpoint = (queryTable) => {
    const datasetCreds = JSON.parse(tableau.connectionData)
    return `https://query.data.world/sql/${datasetCreds.dataset}?authentication=Bearer+${datasetCreds.apiToken}&query=SELECT%20*%20FROM%20%60${queryTable}%60`
  }

  getDatatype = (datatype) => {
    if (schemaMap[datatype]) {
      return schemaMap[datatype]
    } else {
      return tableau.dataTypeEnum.string
    }
  }

  /**
   * Used to verify that the dataset exists and the API key works
   */
  verify = () => {
    return axios.get(this.getApiEndpoint(queryTable))
  }

  getSchema = (callback) => {
    axios.get(this.getApiEndpoint(queryTable)).then((resp) => {
      const datasetTablesResults = resp.data.results.bindings
      const metadata = resp.data.metadata
      const datasetTables = []

      const metadataMap = {}

      metadata.forEach((m, index) => {
        metadataMap[m.name] = resp.data.head.vars[index]
      })
      const {columnIndex, tableName, columnName, columnDatatype} = metadataMap

      for (let i = 0; i < datasetTablesResults.length; i += 1) {
        if (datasetTablesResults[i][columnIndex].value === '1') {
          const activeTable = datasetTablesResults[i][tableName].value
          const datasetCols = []

          for (let j = 0, len = datasetTablesResults.length; j < len; j += 1) {
            if (datasetTablesResults[j][tableName].value === activeTable) {
              const columnId = 'v_' + (parseInt(datasetTablesResults[j][columnIndex].value, 10) - 1)
              datasetCols.push({
                id: columnId,
                alias: datasetTablesResults[j][columnName].value,
                dataType: this.getDatatype(datasetTablesResults[j][columnDatatype].value)
              })
            }
          }

          const datasetTableId = activeTable.replace(/[^A-Z0-9]/ig, '')
          const datasetTable = {
            id: datasetTableId,
            alias: activeTable,
            columns: datasetCols
          }

          datasetTables.push(datasetTable)
        }
      }

      callback(datasetTables)
    }).catch((error) => {
      tableau.log(error)
      tableau.log('There was an error retrieving the schema')
      tableau.abortWithError(error)
    })
  }

  getData = (table, callback) => {
    const filesApiCall = this.getApiEndpoint(table.tableInfo.alias)

    axios.get(filesApiCall).then((resp) => {
      const results = resp.data.results.bindings
      const columnIds = resp.data.head.vars
      const tableData = []

      let i, j


      for (i = 0; i < results.length; i += 1) {
        const jsonData = {}
        for (j = 0; j < columnIds.length; j += 1) {
          const id = columnIds[j]
          if (results[i][id]) {
            if (results[i][columnIds[j]].datatype === 'http://www.w3.org/2001/XMLSchema#boolean') {
              jsonData[id] = results[i][columnIds[j]].value === 'true'
            } else {
              jsonData[id] = results[i][columnIds[j]].value
            }
          }
        }
        tableData.push(jsonData)
      }

      table.appendRows(tableData)
      callback()
    }).catch(function (error) {
      tableau.log(error)
      tableau.log('There was an error retrieving the data')
      tableau.abortWithError(error)
    })
  }

  setConnectionData = (dataset, apiToken) => {
    tableau.connectionData = JSON.stringify({dataset, apiToken})
    tableau.connectionName = dataset
  }
}
