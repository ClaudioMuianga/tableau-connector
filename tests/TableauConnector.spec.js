jest.mock('axios')

import './util'
import TableauConnector from '../src/TableauConnector'
import {schemaData, tableData} from './apiResponseData'
import axios from 'axios'

it('Initializes the tableau connector correctly', () => {
  const connector = new TableauConnector()
  expect(global.tableau.makeConnector.mock.calls.length).toBe(1)
  expect(global.tableau.registerConnector.mock.calls.length).toBe(1)

  const internalConnector = connector.connector
  expect(typeof internalConnector.getSchema).toBe('function')
  expect(typeof internalConnector.getData).toBe('function')
})

it('Returns the correct API endpoint', () => {
  const connector = new TableauConnector()
  connector.setConnectionData('test/1234', 'apitoken-test')
  expect(connector.getApiEndpoint('table-name-test'))
    .toBe('https://query.data.world/sql/test/1234?authentication=Bearer+apitoken-test&query=SELECT%20*%20FROM%20%60table-name-test%60')
})

it('Returns default tableauschema type', () => {
  const connector = new TableauConnector()
  expect(connector.getDatatype('randommissingdatatype')).toBe('string')
})

it('Formats the schema correctly', (done) => {
  axios.__setMockResponse(schemaData)
  const connector = new TableauConnector()
  connector.getSchema((schemaData) => {
    expect(schemaData).toHaveLength(3)
    expect(schemaData).toMatchSnapshot()

    done()
  })
})

it('formats the table correctly', (done) => {
  axios.__setMockResponse(tableData)
  const connector = new TableauConnector()
  const table = {
    tableInfo: {
      alias: 'test'
    },
    appendRows: jest.fn()
  }
  connector.getData(table, () => {
    expect(table.appendRows.mock.calls.length).toBe(1)
    expect(table.appendRows.mock.calls[0][0]).toEqual([
      {
        v_0: '2016-12-14',
        v_1: 'Test test test'
      },
      {
        v_0: '2016-12-14',
        v_1: 'Test test'
      }
    ])
    done()
  })
})