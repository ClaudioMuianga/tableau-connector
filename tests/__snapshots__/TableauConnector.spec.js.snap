exports[`test Formats the schema correctly 1`] = `
Array [
  Object {
    "alias": "AnIntrotodata.worldDatasetChangeLog-Sheet1",
    "columns": Array [
      Object {
        "alias": "Date",
        "dataType": "date",
        "id": "v_0",
      },
      Object {
        "alias": "Change",
        "dataType": "string",
        "id": "v_1",
      },
    ],
    "id": "AnIntrotodataworldDatasetChangeLogSheet1",
  },
  Object {
    "alias": "DataDotWorldBBallStats",
    "columns": Array [
      Object {
        "alias": "Name",
        "dataType": "string",
        "id": "v_0",
      },
      Object {
        "alias": "PointsPerGame",
        "dataType": "float",
        "id": "v_1",
      },
      Object {
        "alias": "AssistsPerGame",
        "dataType": "float",
        "id": "v_2",
      },
    ],
    "id": "DataDotWorldBBallStats",
  },
  Object {
    "alias": "DataDotWorldBBallTeam",
    "columns": Array [
      Object {
        "alias": "Name",
        "dataType": "string",
        "id": "v_0",
      },
      Object {
        "alias": "Height",
        "dataType": "string",
        "id": "v_1",
      },
      Object {
        "alias": "Handedness",
        "dataType": "string",
        "id": "v_2",
      },
    ],
    "id": "DataDotWorldBBallTeam",
  },
]
`;
